HULKs/hulk
# Behavior simulator

## Introduced Changes

Adds a new simulator just for behavior.
See #36 for details.

Fixes #36

Aids in fixing #37 and other rule changes.

## ToDo / Known Issues

- [x] Execute nodes
- [x] Store database for every frame
- [x] Make database available via communication
- [x] Simulate node inputs
  - [x] ball position
  - [x] game controller state
  - [x] robot to field
  - [x] cycle time
  - [x] incoming messages
  - [x] filtered game state
  - [x] primary state
  - [x] role
- [x] Move to `tools/`
- [x] Clean up twix hacks
- [x] Allow choosing the "active" robot via communication
- [x] Scenario files
- [x] Forward messages between robots
- [x] Argument parsing

## Ideas for Next Iterations (Not This PR)

In no particular order

- [ ] Rerun simulation when
  - [ ] scenario file changes
  - [ ] changing parameters via communication
- [ ] Simulate node inputs
  - [ ] obstacles (just run obstacle filter node?)
  - [ ] fall state
  - [ ] penalty shot direction
  - [ ] has ground contact
  - [ ] Ball "physics"
- [ ] Limit robot's field of view
- [ ] Add configurable noise to
  - [ ] ball position
  - [ ] network delay
  - [ ] movement (? idk if this makes any meaningful difference)
- [ ] Apply SPL rules
- [ ] HULK standard library in Lua
- [ ] Run simulator in CI
- [ ] Automatically generate behavior cycler code.

## How to Test

Use this command to simulate a game and start a communication server that you can connect to with twix.

```sh
cargo run --profile incremental --manifest-path tools/behavior_simulator/Cargo.toml <path to scenario lua file>
```

Open a "Behavior Simulator" type panel, to control the timeline. All other twix panels should function normally, although only the `Control` cycler exists. Subscribing outputs from any other cycler will result in an error.

The behavior simulator map layer should show all playing robots. While all other panels and layers should show the values for the currently selected robot. This selection is done via the behavior simulator panel.

A sample scenario file can be found at `tests/behavior/demonstration.lua`.
Note that the example uses the `inspect` lua library. Please install it via your system's package manager or a lua package manager like luarocks.
